### PR TITLE
Resolve module specifiers for auto imports in completion list (in incomplete chunks)

### LINKED ISSUE #42005: Improve "suggestions show path" for "node_modules"
## Search Terms

TypeScript suggestions show path
TypeScript can provide completions that also add an import statement. However, when there are multiple symbols with the same name, it is hard to pick the right completion. This release makes this simpler because paths of auto-import completions are shown with the label.

## Suggestion

As description above. I think the path should be show as `import` form, not show path like 'node_modules/xxx/index' but 'xxx'
And sort it as the "show fixes" menu I post below.

## Use Cases
Here is the current result

图片
It is not very easy to choose which one to import `Table`, because the path is too long and very same. (Actually I want the 3rd one, it is 'antd')

----
Here is the "show fixes" menu
图片
It is much more clear.

----

Here is the "show fixes" menu, when I already import something from 'antd'
图片

## Examples

no example.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #42005

## Summary

- Before May 2020: auto imports showed no secondary label in completion lists. They had a different icon, but you had to focus them to really see that they were auto imports at all, much less where they were going to come from.
- May 2020–present: auto imports showed a workspace-relative path to _a_ module that exports the symbol. If the symbol was re-exported from multiple places, the path shown may not be the path actually imported from. Also, path mappings, node_modules paths, index file paths, etc. were not taken into account—it was literally just a relative path, not a module specifier.
- This PR: it's too slow to resolve module specifiers unconditionally for every auto import in an arbitrarily long completion list, so when you first trigger completions, we compute up to 100 module specifiers, and try the module specifier cache up to an additional 1000 times, and fill them into the completion list. Entries beyond what we can get module specifiers for will continue to show up with the old module path (though I’m open to tweaking their display in a way that doesn’t impact perf in another PR). Then, as you continue to type characters, if any entries were left with unresolved module specifiers VS Code will re-request completions. We pull the previous response from a cache and fill in another 100/1000 module specifiers, but only for entries that still match the identifier you’re typing—a large chunk will get filtered out with every request. These continuation requests are very fast—58 ms max in my tests—and they usually fully complete the list within one or two requests.

## Screen capture 🎥 

You can see a smattering of fully resolved module specifiers in the initial list here, but some, like `changeDepenciesStateTo0` still say `node_modules`. But once I’ve type enough characters to find it without scrolling, its label has updated to be its full module specifier.

https://user-images.githubusercontent.com/3277153/123308652-86dc8700-d4e9-11eb-862a-34429879fb41.mp4

It occurred to me while making this recording that it would probably be better to try resolving module specifiers in alphabetical order (instead of module traversal order), so all the resolved ones would be more toward the top of the list. However, this may not matter as much as it feels like it would since VS Code by default snaps you to names you recently completed elsewhere. So in the real world, you might be snapped to an essentially random spot in the list instead of near the top, depending on your settings. But I’ll think about whether the order can be smarter in some way as a future optimization.

## Significant related changes

- Ambient modules don’t count against the resolution limit—they always get fully resolved in the first go now, because they’re very cheap to resolve and we notice that they’re ambient while putting them in the export map cache. This applies even to older clients who otherwise can’t use this feature because they don’t signal that they can support incomplete responses/retriggerings.

- As part of the way we can keep the performance cost of this down, the fuzzy matching logic for auto imports has changed from “the symbol name contains the typed characters, in order, anywhere in it” to a slightly more restrictive version, akin to what was used for import statement completions in #43149. I can’t do better to explain it than I did in the doc comment:

   > True if the first character of `lowercaseCharacters` is the first character of some "word" in `identiferString` (where the string is split into "words" by camelCase and snake_case segments), then if the remaining characters of `lowercaseCharacters` appear, in order, in the rest of `identifierString`.
   >
   > True:
   > 'state' in 'useState'
   > 'sae' in 'useState'
   > 'viable' in 'ENVIRONMENT_VARIABLE'
   >
   > False:
   > 'staet' in 'useState'
   > 'tate' in 'useState'
   > 'ment' in 'ENVIRONMENT_VARIABLE'

## Performance measurements

In a test project containing lots of expensive dependencies (especially `aws-sdk`):

|           | type 'c' | type 'o' | delete word | type 'c' | type 'r' |
|-------|-------:|--------:|-------------:|-------:|--------:|
| 4.3.2 | 405 ms | -           | -                    | 136 ms | -           |
| PR     | 415 ms | 58 ms  | -                     | 220 ms | 20 ms |

I’m not thrilled with 136 → 220 for the cached scenario, but that number is only so low in the first place because of cache improvements that have been made over the last year or two, and it’s still low in absolute terms for a completion list that large. I plan to continue optimizations here as needed, but I’m ok with these numbers considering the value added.

### CODE REVIEW

File: src/services/codefixes/importFixes.ts [hunk 1]
```diff
@@ -321,12 +351,52 @@ namespace ts.codefix {
...
+                            !!ambientModuleName);
+                    });
+                },
+            };
+            if (Debug.isDebugging) {
+                Object.defineProperty(wrapped, "__cache", { get: () => map });
+            }
+            return wrapped;
+
+            function replaceTransientSymbols(info: SymbolExportInfo, checker: TypeChecker) {
```
Comment (andrewbranch) [inline src/services/codefixes/importFixes.ts:391]: Transient symbols should never have been in this cache, but this isn’t new, it’s just moved and renamed. I’m working on a follow-up to restructure this cache to be safer and less kludgy.

File: src/services/completions.ts [hunk 2]
```diff
@@ -3150,4 +3300,58 @@ namespace ts.Completions {
...
+        let prevChar: number | undefined;
+        let characterIndex = 0;
+        const len = identifierString.length;
+        for (let strIndex = 0; strIndex < len; strIndex++) {
+            const strChar = identifierString.charCodeAt(strIndex);
+            const testChar = lowercaseCharacters.charCodeAt(characterIndex);
+            if (strChar === testChar || strChar === toUpperCharCode(testChar)) {
+                matchedFirstCharacter ||=
+                    prevChar === undefined || // Beginning of word
+                    CharacterCodes.a <= prevChar && prevChar <= CharacterCodes.z && CharacterCodes.A <= strChar && strChar <= CharacterCodes.Z || // camelCase transition
```
Comment (sandersn) [inline src/services/completions.ts:3335]: future work: maybe use general unicode categories to detect lower -> uppercase transition

**Review by sandersn — APPROVED** [review]
Looks good based on my understanding -- thanks for walking through the code with me.

File: src/services/completions.ts [hunk 3]
```diff
@@ -3150,4 +3300,58 @@ namespace ts.Completions {
...
+
+    /**
+     * True if the first character of `lowercaseCharacters` is the first character
+     * of some "word" in `identiferString` (where the string is split into "words"
+     * by camelCase and snake_case segments), then if the remaining characters of
+     * `lowercaseCharacters` appear, in order, in the rest of `identifierString`.
+     *
+     * True:
+     * 'state' in 'useState'
+     * 'sae' in 'useState'
```
Comment (amcasey) [inline src/services/completions.ts:3312]: Could have been "sate". 😛

File: src/services/completions.ts (hunk 2 above)
Comment (amcasey) [reply src/services/completions.ts:3335]: This seems important, e.g. for languages with accents (I assume languages without lettercase use snake case?).

File: src/services/completions.ts [hunk 4]
```diff
@@ -3150,4 +3300,58 @@ namespace ts.Completions {
...
+                }
+            }
+            prevChar = strChar;
+        }
+
+        // Did not find all characters
+        return false;
+    }
+
+    function toUpperCharCode(charCode: number) {
```
Comment (amcasey) [inline src/services/completions.ts:3351]: I think we usually use toLower, though I don't know whether that was to improve the behavior.

**Review by amcasey — COMMENTED** [review]
The protocol change LGTM and the explanation makes sense.

**Pull request merged**